### DISCUSSION TITLE
How can I add custom analytics with auto-removal of UTM parameters?

### DISCUSSION THREAD

**Discussion by matkoch** [facebook/docusaurus#6671] [Q&A] [answered]
I'm using another analytics service (not GA) called Fathom. How do I include this?

Also, how can I remove UTM parameters after a short while to allow visitors to copy the link without forwarding the original parameters (to let it count as organic traffic)?

**Comment by matkoch** [comment, marked as answer]
I adapted my files as follows. Note that Docusaurus (as of now), triggers `onRouteUpdate` too early, so we have to manually trigger analytics.

`docusaurus.config.js`
```
const config = {
  scripts: [
    {
      src: 'https://cdn.usefathom.com/script.js',
      defer: true,
      'data-site': 'XXXX',
      'data-spa': 'auto',
    },
  ],
  clientModules: [
    require.resolve('./analytics.js'),
  ],
}
```

`src/theme/Root.js` (currently, it doesn't allow swizzling, just create it):
```
import React from 'react';
import ExecutionEnvironment from "@docusaurus/ExecutionEnvironment";

function Root({children}) {
    if (ExecutionEnvironment.canUseDOM) {
        const parts = window.location.href.split("?");
        if (parts.length > 1) {
            setTimeout(() => window.history.replaceState({}, document.title, parts[0]), 4000);
        }
    }

    return <>{children}</>;
}

export default Root;
```

`analytics.js`
```
import ExecutionEnvironment from '@docusaurus/ExecutionEnvironment'

export default (function () {
    if (!ExecutionEnvironment.canUseDOM) {
        return null
    }

    return {
        onRouteUpdate({ location }) {
            setTimeout(() => fathom.trackPageview(), 3000);
        }
    }
})()
```

As of now, a 2.0-beta build should be used. It seems 1.0 has a bug with canonical meta data.

**Reply by matkoch** [reply]
Thanks @Josh-Cena! Added my solution here to make it more discoverable.

**Reply by slorber** [reply]
yes, onRouteUpdate fires too early

https://github.com/facebook/docusaurus/issues/3399

As a workaround you can also add a component in Root and call your analytics in `useEffect()` when some value from `useLocation` changes

Note this import looks bad to me, we may prevent you from doing this in the future:

> `@docusaurus/core/lib/client/exports/ExecutionEnvironment`

**Reply by matkoch** [reply]
What's the alternative?

**Reply by Josh-Cena** [reply]
[`@docusaurus/ExecutionEnvironment`](https://docusaurus.io/docs/docusaurus-core#executionenvironment)

**Reply by matkoch** [reply]
Strange though that this is not recognized by my IDE...

**Reply by Josh-Cena** [reply]
@slorber w.r.t. the `onRouteUpdate` firing too early, should we simply use `componentDidUpdate`, and use a ref that's written to `true` when the update is triggered by a route change released from `shouldComponentUpdate`?

**Reply by Josh-Cena** [reply]
@matkoch That's a webpack alias. You probably need to use `@docusaurus/module-type-aliases` to get it recognized by your editor. See https://docusaurus.io/docs/typescript-support

**Reply by slorber** [reply]
@Josh-Cena I'd rather not do a change in this old code before we refactor it. 

Eventually, we can already migrate this lifecycle call in a child component in `useEffect`: this way the logic will stay exactly the same after we refactor that navigation component.

**Reply by Josh-Cena** [reply]
So, the `PendingNavigation` will only be responsible for intercepting the route transition, but the actual lifecycle dispatching will be in a child layer? Sounds good

**Reply by slorber** [reply]
Yes that's the idea, useEffect should be good enough to fire that event at the right time IMHO, but it requires to define properly what is a route change too, cf https://github.com/facebook/docusaurus/issues/3399#issuecomment-1024090449)
